distributed/dashboard/components/scheduler.py

```diff
@@ -1720,6 +1724,214 @@ def __del__(self):
         self.scheduler.remove_plugin(self.layout)
 
 
+##########Task Group Graph #########
+class TGroupGraph(DashboardComponent):
+    """
+    Task Group Graph
+    """
+
+    def __init__(self, scheduler, **kwargs):
+        self.scheduler = scheduler
+        self.x = {}
+        self.y = {}
+        self.collision = {}
+        self.y_next = 0
+        self.nstart = []
+        self.nend = []
+        self.tg_stack = []
+
+        self.node_source = ColumnDataSource(
+            {
+                "x": [],
+                "y": [],
+                "name": [],
+                "name_short": [],
+                "tot_tasks": [],
+                "colors": [],
+            }
+        )
+
+        self.arrow_source = ColumnDataSource({"xs": [], "ys": [], "xe": [], "ye": []})
+
+        self.root = figure(title="Task Groups Graph", **kwargs)
+        self.subtitle = Title(text=" ", text_font_style="italic")
+        self.root.add_layout(self.subtitle, "above")
+
+        # Let's start with squares we, will modify later to progress bars
+        rect = self.root.square(
+            x="x",
+            y="y",
+            size=30,
+            color="colors",
+            fill_alpha=0.5,
+            source=self.node_source,
+        )
+
+        self.arrows = Arrow(
+            end=VeeHead(size=10),
+            line_color="black",
+            line_alpha=0.5,
+            line_width=1,
+            x_start="xs",
+            y_start="ys",
+            x_end="xe",
+            y_end="ye",
+            source=self.arrow_source,
+        )
+        self.root.add_layout(self.arrows)
+
+        self.labels = LabelSet(
+            x="x",
+            y="y",
+            text="name_short",
+            source=self.node_source,
+            background_fill_color=None,
+        )  #  We probably need an offset like x_offset=-1, also need to chop the name somehow
+        self.root.add_layout(self.labels)
+
+        self.root.xgrid.grid_line_color = None
+        self.root.ygrid.grid_line_color = None
+
+        hover = HoverTool(
+            point_policy="follow_mouse",
+            tooltips=[("tg", "@name"), ("num_task", "@tot_tasks")],
+            renderers=[rect],
+        )
+        self.root.add_tools(hover)
+
+    @without_property_validation
+    def update(self):
```

Comment:
We may want to avoid updating things if there have not been any new task groups.  This may significantly cut down on recomputation costs.  Once we start including things like progressbars then we'll want to only update layout when task groups change, but keep updating visual attributes every time tasks change.